### ISSUE TITLE
New dsl, step 3/3: migrate tag-specific prop types from Html to Dom_html

### ISSUE THREAD

**Issue by jchavarri** [ml-in-barcelona/jsoo-react#136]
After original api implementation in https://github.com/ml-in-barcelona/jsoo-react/pull/119 and update of ppx to support it in https://github.com/ml-in-barcelona/jsoo-react/pull/128, the last step of this new api consists on migrating the "specs" implemented with types in ppx [`Html` module](https://github.com/ml-in-barcelona/jsoo-react/blob/9e86c9f4aeb5050636a2fc36c9f0cf45f5af6b20/ppx/html.ml) to the `Prop` module in `Dom_html`. The goal being to make impossible to pass wrong attributes to e.g. a `div`.

I am not sure about the best way to implement this, probably could use [an approach similar to `bs-css`](https://github.com/giraud/bs-css/blob/49cb3cc54dcd0efd33a934ec45cd8a3354fd5bb7/bs-css/src/Css_AtomicTypes.rei#L64-L83), with polyvariants + subtyping to reflect inheritance or composition based on smaller subsets of props, like [here](https://github.com/ml-in-barcelona/jsoo-react/blob/9e86c9f4aeb5050636a2fc36c9f0cf45f5af6b20/ppx/html.ml#L1649).

**Comment by glennsl** [comment]
This sounds very complicated, and I'm not really sure it's worth it. I can't remember a single instance of having used the wrong attributes for a tag, which might also be because the consequence of doing so is entirely benign. In almost all cases it would just not do anything. You might get a warning in the browser console, and HTML validation would fail, but that's about it.

There is one feature that I think would make it worthwhile, however: The ability to auto-complete just the valid props for a tag. Unfortunately I have no idea how to even approach a solution for that problem.

**Comment by glennsl** [comment]
I also think it was actually me who came up with the scheme of using polyvariants to type CSS. At least I did a proper go at it with [bs-typed-css](https://github.com/glennsl/bs-typed-css-core/blob/master/src/CssCore.re), and came to two conclusions from that: 1) it's very hard to do, and 2) there's very little benefit. Today I just use strings as values, and have very few issues with that approach.

It is a very fun challenge to try to come up with a scheme to make such a complicated system well-typed. So please don't let this discourage you from experimenting with it. I'm just not convinced this is all that valuable for production code.

**Comment by jchavarri** [comment]
Yeah I think these points are worth considering. What you mention makes me think about tyxml, which also left me (as a user) with similar feeling: the time spent trying to debug type errors was > time saved by wrong elements being created, just because the latter happens so rarely.

In general, considering how few resources jsoo-react has :) I tend to pick the simplicity path over the complexity one, unless the complexity is absolutely needed.

On the other hand, to make full argument, these constraints are very unlikely to change over time, and also, regarding how complex the implementation is, users will not perceive these complexity, because the api design would not change.

@davesnx what do you think?

**Comment by davesnx** [comment]
I fully agree, I personally prefer to keep the API very unsafe in those cases. bs-css makes it look easy for simple cases, but for using more advanced CSS features is totally a nightmare, is one of the reasons that I started styled-ppx in the first place!
